File: toolbox/lib/acme_aabb.m
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
%                                                                     %
% The ACME project                                                    %
%                                                                     %
% Copyright (c) 2020-2021, Davide Stocco and Enrico Bertolazzi.       %
%                                                                     %
% The ACME project and its components are supplied under the terms of %
% the open source BSD 2-Clause License. The contents of the ACME      %
% project and its components may not be copied or disclosed except in %
% accordance with the terms of the BSD 2-Clause License.              %
%                                                                     %
% URL: https://opensource.org/licenses/BSD-2-Clause                   %
%                                                                     %
%    Davide Stocco                                                    %
%    Department of Industrial Engineering                             %
%    University of Trento                                             %
%    e-mail: davide.stocco@unitn.it                                   %
%                                                                     %
%    Enrico Bertolazzi                                                %
%    Department of Industrial Engineering                             %
%    University of Trento                                             %
%    e-mail: enrico.bertolazzi@unitn.it                               %
%                                                                     %
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %

classdef acme_aabb < acme_entity
  %
  % Class container for ACME disk object
  %
  methods
    function self = acme_aabb( varargin )
      % Create a new C++ pointer to disk object instance
      self.objectHandle = mex_aabb( 'new', varargin{:} );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function delete( self )
      % Create C++ pointer to disk object instance
      mex_aabb( 'delete', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMin( self )
      % Get minimum aabb as ACME point
      out = acme_point();
      out.copyByHandle( mex_aabb( 'getMin', self.objectHandle ) );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMinX( self )
      % Get aabb minimum point X value
      out = mex_aabb( 'getMinX', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMinY( self )
      % Get aabb minimum point X value
      out = mex_aabb( 'getMinY', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMinZ( self )
      % Get aabb minimum point Z value
      out = mex_aabb( 'getMinZ', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMax( self )
      % Get maximum aabb as ACME point
      out = acme_point();
      out.copyByHandle( mex_aabb( 'getMax', self.objectHandle ) );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMaxX( self )
      % Get aabb maximum point X value
      out = mex_aabb( 'getMaxX', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMaxY( self )
      % Get aabb maximum point X value
      out = mex_aabb( 'getMaxY', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = getMaxZ( self )
      % Get aabb maximum point Z value
      out = mex_aabb( 'getMaxZ', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMin( self, other_obj )
      % Set minimum aabb as ACME point
      if (other_obj.type() == "point")
        mex_aabb( 'setMin', self.objectHandle, other_obj.objectHandle );
      else
        error('mex_aabb::setMin(): other_obj must be an ACME point object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMinX( self, value )
      % Set aabb minimum point X value
      mex_aabb( 'setMinX', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMinY( self, value )
      % Set aabb minimum point X value
      mex_aabb( 'setMinY', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMinZ( self, value )
      % Set aabb minimum point Z value
      mex_aabb( 'setMinZ', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMax( self, other_obj )
      % Set maximum aabb as ACME point
      if (other_obj.type() == "point")
        mex_aabb( 'setMax', self.objectHandle, other_obj.objectHandle );
      else
        error('mex_aabb::setMax(): other_obj must be an ACME point object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMaxX( self, value )
      % Set aabb maximum point X value
      mex_aabb( 'setMaxX', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMaxY( self, value )
      % Set aabb maximum point X value
      mex_aabb( 'setMaxY', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function setMaxZ( self )
      % Set aabb maximum point Z value
      mex_aabb( 'setMaxZ', self.objectHandle, value );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function copy( self, other_obj )
      % Copy aabb object from another aabb
      if (other_obj.type() == "aabb")
        mex_aabb( 'copy', self.objectHandle, other_obj.objectHandle );
      else
        error('mex_aabb::copy(): other_obj must be an ACME aabb object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function copyByHandle( self, handle )
      % Copy aabb object from another aabb handle
      mex_aabb( 'copy', self.objectHandle, handle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function translate( self, other_obj )
     % Translate aabb by vector
      mex_aabb( 'translate', self.objectHandle, other_obj.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = isInside( self, other_obj )
      % Check if ACME point is inside the aabb 
      if (other_obj.type() == "point")
        out = mex_aabb( 'isInside', self.objectHandle, other_obj.objectHandle );
      else
         error('mex_aabb::isInside(): other_obj must be an ACME point object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = isDegenerated( self )
      % Check if aabb is degenerated
      out = mex_aabb( 'isDegenerated', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = isApprox( self, other_obj )
      % Check if aabbs are approximatively equal
      if (other_obj.type() == "aabb") 
        out = mex_aabb( 'isApprox', self.objectHandle, other_obj.objectHandle );
      else
         error('mex_aabb::isApprox(): other_obj must be an ACME aabb object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = id( self )
      % Return aabb id
      out = mex_aabb( 'id', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = pos( self )
     % Return aabb position
     out = mex_aabb( 'pos', self.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = intersection( self, other_obj )
      if (other_obj.type() == "aabb") 
        out = mex_aabb( 'intersection', self.objectHandle, other_obj.objectHandle );
      else
        error('mex_aabb::intersection(): other_obj must be an ACME aabb object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = intersects( self, other_obj )
      if (other_obj.type() == "aabb") 
        out = mex_aabb( 'intersects', self.objectHandle, other_obj.objectHandle );
      else
        error('mex_aabb::intersects(): other_obj must be an ACME aabb object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function disp( self )
      % Display object data
      disp( [ self.getMin().get(),  self.getMax().get() ] );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function plot( self, figure_name, color )
      % Plot disk object
      figure_name;
      hold on;
      min = self.getMin().get();
      max = self.getMax().get();
      plot3( [min(1), max(1)], ... % 01
             [min(2), min(2)], ...
             [min(3), min(3)], ...
             'Color', color)
      plot3( [min(1), min(1)], ... % 02
             [min(2), max(2)], ...
             [min(3), min(3)], ...
             'Color', color)
      plot3( [min(1), min(1)], ... % 03
             [min(2), min(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [max(1), min(1)], ... % 04
             [max(2), max(2)], ...
             [max(3), max(3)], ...
             'Color', color)
      plot3( [max(1), max(1)], ... % 05
             [max(2), min(2)], ...
             [max(3), max(3)], ...
             'Color', color)
      plot3( [max(1), max(1)], ... % 06
             [max(2), max(2)], ...
             [max(3), min(3)], ...
             'Color', color)
      plot3( [max(1), max(1)], ... % 07
             [min(2), max(2)], ...
             [min(3), min(3)], ...
             'Color', color)
      plot3( [min(1), min(1)], ... % 08
             [min(2), max(2)], ...
             [max(3), max(3)], ...
             'Color', color)
      plot3( [min(1), max(1)], ... % 09
             [min(2), min(2)], ...
             [max(3), max(3)], ...
             'Color', color)
      plot3( [min(1), min(1)], ... % 10
             [max(2), max(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [max(1), max(1)], ... % 11
             [min(2), min(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [min(1), max(1)], ... % 12
             [max(2), max(2)], ...
             [min(3), min(3)], ...
             'Color', color)
      % Cross lines on faces
      plot3( [min(1), max(1)], ... % 13
             [min(2), max(2)], ...
             [min(3), min(3)], ...
             'Color', color)
      plot3( [min(1), max(1)], ... % 14
             [min(2), max(2)], ...
             [max(3), max(3)], ...
             'Color', color)
      plot3( [min(1), min(1)], ... % 15
             [min(2), max(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [max(1), max(1)], ... % 16
             [min(2), max(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [min(1), max(1)], ... % 17
             [min(2), min(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      plot3( [min(1), max(1)], ... % 18
             [max(2), max(2)], ...
             [min(3), max(3)], ...
             'Color', color)
      hold off;
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    function out = type( self )
      % Get object type as string
      out = 'aabb';
    end
  end
end

File: src/acme_aabb.cc
/*
(***********************************************************************)
(*                                                                     *)
(* The ACME project                                                    *)
(*                                                                     *)
(* Copyright (c) 2020-2021, Davide Stocco and Enrico Bertolazzi.       *)
(*                                                                     *)
(* The ACME project and its components are supplied under the terms of *)
(* the open source BSD 2-Clause License. The contents of the ACME      *)
(* project and its components may not be copied or disclosed except in *)
(* accordance with the terms of the BSD 2-Clause License.              *)
(*                                                                     *)
(* URL: https://opensource.org/licenses/BSD-2-Clause                   *)
(*                                                                     *)
(*    Davide Stocco                                                    *)
(*    Department of Industrial Engineering                             *)
(*    University of Trento                                             *)
(*    e-mail: davide.stocco@unitn.it                                   *)
(*                                                                     *)
(*    Enrico Bertolazzi                                                *)
(*    Department of Industrial Engineering                             *)
(*    University of Trento                                             *)
(*    e-mail: enrico.bertolazzi@unitn.it                               *)
(*                                                                     *)
(***********************************************************************)
*/

///
/// file: acme_aabb.cc
///

#include "acme_aabb.hh"

namespace acme
{

  /*\
   |               _     _     
   |    __ _  __ _| |__ | |_     
   |   / _` |/ _` | '_ \| '_ \ 
   |  | (_| | (_| | |_) | |_) |
   |   \__,_|\__,_|_.__/|_.__/ 
   |                           
  \*/

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  aabb &
  aabb::operator=(
      aabb const &input)
  {
    if (this == &input)
    {
      return *this;
    }
    else
    {
      this->_min = input._min;
      this->_max = input._max;
      this->_id = input._id;
      this->_ipos = input._ipos;
      return *this;
    }
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::clear(
      void)
  {
    this->_min = NaN_point;
    this->_max = NaN_point;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::isApprox(
      aabb const &input,
      real tolerance)
      const
  {
    return this->_min.isApprox(input._min, tolerance) &&
           this->_max.isApprox(input._max, tolerance);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::checkMaxMin(void)
      const
  {
    return this->_max.x() >= this->_min.x() &&
           this->_max.y() >= this->_min.y() &&
           this->_max.z() >= this->_min.z();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::updateMaxMin(void)
  {
    bool output = true;
    real point_max_tmp, point_min_tmp;
    for (integer i = 0; i < 3; ++i)
    {
      point_max_tmp = this->_max[i];
      point_min_tmp = this->_min[i];
      if (point_max_tmp < point_min_tmp)
      {
        this->_max[i] = point_min_tmp;
        this->_min[i] = point_max_tmp;
        output = false;
      }
    }
    return output;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point const &
  aabb::min(void)
      const
  {
    return this->_min;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::minX(void)
      const
  {
    return this->_min.x();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::minY(void)
      const
  {
    return this->_min.y();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::minZ(void)
      const
  {
    return this->_min.z();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::min(
      integer i)
      const
  {
    return this->_min[i];
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::min(
      point const &input)
  {
    this->_min = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::min(
      real x,
      real y,
      real z)
  {
    this->_min.x() = x;
    this->_min.y() = y;
    this->_min.z() = z;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::minX(
      real input)
  {
    this->_min.x() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::minY(
      real input)
  {
    this->_min.y() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::minZ(
      real input)
  {
    this->_min.z() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::min(
      integer i,
      real input)
  {
    this->_min[i] = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point const &
  aabb::max(void)
      const
  {
    return this->_max;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::maxX(void)
      const
  {
    return this->_max.x();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::maxY(void)
      const
  {
    return this->_max.y();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::maxZ(void)
      const
  {
    return this->_max.z();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::max(
      integer i)
      const
  {
    return this->_max[i];
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::max(
      point const &input)
  {
    this->_max = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::max(
      real x,
      real y,
      real z)
  {
    this->_max.x() = x;
    this->_max.y() = y;
    this->_max.z() = z;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::maxX(
      real input)
  {
    this->_max.x() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::maxY(
      real input)
  {
    this->_max.y() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::maxZ(
      real input)
  {
    this->_max.z() = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::max(
      integer i,
      real input)
  {
    this->_max[i] = input;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::intersects(
      aabb const &input)
      const
  {
    return (this->_min.x() <= input._max.x() && this->_max.x() >= input._min.x()) &&
           (this->_min.y() <= input._max.y() && this->_max.y() >= input._min.y()) &&
           (this->_min.z() <= input._max.z() && this->_max.z() >= input._min.z());
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::merged(
      std::vector<aabb::ptr> const &boxes)
  {
    if (boxes.empty())
    {
      this->_min = Zeros_point;
      this->_max = Zeros_point;
    }
    else
    {
      std::vector<aabb::ptr>::const_iterator it = boxes.begin();

      this->_min = (*it)->_min;
      this->_max = (*it)->_max;

      for (++it; it != boxes.end(); ++it)
      {
        aabb const &cur_box = **it;
        if (cur_box._min.x() < this->_min.x())
          this->_min.x() = cur_box._min.x();
        if (cur_box._min.y() < this->_min.y())
          this->_min.y() = cur_box._min.y();
        if (cur_box._min.z() < this->_min.z())
          this->_min.z() = cur_box._min.z();
        if (cur_box._max.x() > this->_max.x())
          this->_max.x() = cur_box._max.x();
        if (cur_box._max.y() > this->_max.y())
          this->_max.y() = cur_box._max.y();
        if (cur_box._max.z() > this->_max.z())
          this->_max.z() = cur_box._max.z();
      }
    }
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::centerDistance(
      point const &query_point)
      const
  {
    point center((this->_max + this->_min) / 2);
    point point_max_centered(this->_max - center);
    point point_centered(query_point - center);
    real x_scale = acme::abs(real(1.0) / point_max_centered.x());
    real y_scale = acme::abs(real(1.0) / point_max_centered.y());
    real z_scale = acme::abs(real(1.0) / point_max_centered.z());
    real dx = acme::max(real(0.0), acme::abs(point_centered.x()) * x_scale - real(1.0)) / x_scale;
    real dy = acme::max(real(0.0), acme::abs(point_centered.y()) * y_scale - real(1.0)) / y_scale;
    real dz = acme::max(real(0.0), acme::abs(point_centered.z()) * z_scale - real(1.0)) / z_scale;
    return acme::sqrt(dx * dx + dy * dy + dz * dz);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real
  aabb::exteriorDistance(
      point const &query_point)
      const
  {
    real dx = acme::max(acme::abs(query_point.x() - this->_min.x()), acme::abs(query_point.x() - this->_max.x()));
    real dy = acme::max(acme::abs(query_point.y() - this->_min.y()), acme::abs(query_point.y() - this->_max.y()));
    real dz = acme::max(acme::abs(query_point.z() - this->_min.z()), acme::abs(query_point.z() - this->_max.z()));
    return acme::sqrt(dx * dx + dy * dy + dz * dz);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::clamp(
      point const &point_0,
      point const &point_1,
      point const &point_2)
  {
    this->_min.x() = acme::min(point_0.x(), point_1.x(), point_2.x());
    this->_min.y() = acme::min(point_0.y(), point_1.y(), point_2.y());
    this->_min.z() = acme::min(point_0.z(), point_1.z(), point_2.z());
    this->_max.x() = acme::max(point_0.x(), point_1.x(), point_2.x());
    this->_max.y() = acme::max(point_0.y(), point_1.y(), point_2.y());
    this->_max.z() = acme::max(point_0.z(), point_1.z(), point_2.z());
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::clamp(
      point const points[3])
  {
    this->_min.x() = acme::min(points[0].x(), points[1].x(), points[2].x());
    this->_min.y() = acme::min(points[0].y(), points[1].y(), points[2].y());
    this->_min.z() = acme::min(points[0].z(), points[1].z(), points[2].z());
    this->_max.x() = acme::max(points[0].x(), points[1].x(), points[2].x());
    this->_max.y() = acme::max(points[0].y(), points[1].y(), points[2].y());
    this->_max.z() = acme::max(points[0].z(), points[1].z(), points[2].z());
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  integer const &
  aabb::id(void)
      const
  {
    return this->_id;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  integer const &
  aabb::pos(void)
      const
  {
    return this->_ipos;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::translate(
      vec3 const &input)
  {
    this->_min = input + this->_min;
    this->_max = input + this->_max;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  aabb::transform(
      affine const &matrix)
  {
    this->_min.transform(matrix);
    this->_max.transform(matrix);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::isInside(
      point const &query_point,
      real tolerance)
      const
  {
    return this->_min.x() <= query_point.x() &&
           this->_min.y() <= query_point.y() &&
           this->_min.z() <= query_point.z() &&
           this->_max.x() >= query_point.x() &&
           this->_max.y() >= query_point.y() &&
           this->_max.z() >= query_point.z();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  aabb::isDegenerated(
      real tolerance)
      const
  {
    return this->_min.isApprox(this->_max, tolerance) &&
           this->checkMaxMin();
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

} // namespace acme

///
/// eof: acme_aabb.cc
///